src/bitmap/backend/atomic_bitmap_mmap.rs

```diff
@@ -0,0 +1,608 @@
+        }
+    }
+
+    fn dirty_at(&self, offset: usize) -> bool {
+        let inner = self.inner.read().unwrap();
+        let Some(bitmap) = inner.as_ref() else {
+            return false;
+        };
+
+        bitmap.dirty_at(self.base_address.wrapping_add(offset))
+    }
+
+    fn slice_at(&self, offset: usize) -> <Self as WithBitmapSlice>::S {
+        Self {
+            inner: self.inner.clone(),
+            base_address: self.base_address.wrapping_add(offset),
+        }
+    }
+}
+
+impl BitmapReplace for BitmapRegion {
+    type InnerBitmap = AtomicBitmapMmap;
+
+    fn replace(&self, bitmap: AtomicBitmapMmap) {
+        let mut inner = self.inner.write().unwrap();
+        inner.replace(bitmap);
+    }
+}
+
+impl BitmapSlice for BitmapRegion {}
+
+impl<'a> WithBitmapSlice<'a> for BitmapRegion {
+    type S = Self;
+}
+
+impl NewBitmap for BitmapRegion {
+    fn with_len(_len: usize) -> Self {
+        Self::default()
+    }
+}
+
+/// `AtomicBitmapMmap` implements a simple memory-mapped bitmap on the page level with test
+/// and set operations. The main use case is to support live migration on vhost-user backends
+/// (see `VHOST_USER_PROTOCOL_F_LOG_SHMFD` and `VHOST_USER_SET_LOG_BASE` in the vhost-user protocol
+/// specification). It uses a fixed memory page size of `LOG_PAGE_SIZE` bytes, so it converts
+/// addresses to page numbers before setting or clearing the bits.
+#[derive(Debug)]
+pub struct AtomicBitmapMmap {
+    logmem: MmapLogReg,
+    pages_before_region: usize, // Number of pages to ignore from the start of the bitmap
+    number_of_pages: usize,     // Number of total pages indexed in the bitmap for this region
+}
+
+// `AtomicBitmapMmap` implements a simple bitmap, it is page-size aware and relative
+// to a memory region. It  handling the `log` memory mapped area. Each page is indexed
+// inside a block of `LOG_BLOCK_SIZE` bits, so even if the bitmap starts at the beginning of
+// the mapped area, the memory region does not necessarily have to start at the beginning of
+// that block.
+// Note: we don't implement `Bitmap` because we cannot implement `slice_at()`
+impl AtomicBitmapMmap {
+    // Sets the memory range as dirty. The `offset` is relative to the memory region,
+    // so an offset of `0` references the start of the memory region. Any attempt to
+    // access beyond the end of the bitmap are simply ignored.
+    fn mark_dirty(&self, offset: usize, len: usize) {
+        if len == 0 {
+            return;
+        }
+
+        let first_page = page_number(offset);
+        let last_page = page_number(offset.saturating_add(len - 1));
+        for page in first_page..=last_page {
+            if page >= self.number_of_pages {
+                break; // ignore out of bound access
+            }
+
+            // get the absolute page number
+            let page = self.pages_before_region + page;
+            self.logmem[page_block(page)].fetch_or(1 << page_bit(page), Ordering::SeqCst);
```

Comment:
To be honest, I don't remember using SeqCst (*), I was sure I used Relaxed (maybe I chenged using the autocomplete of my editor and I didn't noticed). Anyway, yes I think Relaxed is enough, thanks.

I want to reuse AtomicBitmap code, that's part of the plan, but it requires more changes that I was trying to avoid.

(*) For some reason, qemu uses SeqCst. I'm not a fan, I think the seqCst model it's more complicated than Relaxed